kernel/src/transaction.rs

```diff
@@ -0,0 +1,543 @@
+/// // stage table changes (right now only commit info)
+/// txn.commit_info(Box::new(ArrowEngineData::new(engine_commit_info)));
+/// // commit! (consume the transaction)
+/// txn.commit(&engine)?;
+/// ```
+pub struct Transaction {
+    read_snapshot: Arc<Snapshot>,
+    operation: Option<String>,
+    commit_info: Option<Arc<dyn EngineData>>,
+}
+
+impl std::fmt::Debug for Transaction {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&format!(
+            "Transaction {{ read_snapshot version: {}, commit_info: {} }}",
+            self.read_snapshot.version(),
+            self.commit_info.is_some()
+        ))
+    }
+}
+
+impl Transaction {
+    /// Create a new transaction from a snapshot. The snapshot will be used to read the current
+    /// state of the table (e.g. to read the current version).
+    ///
+    /// Instead of using this API, the more typical (user-facing) API is
+    /// [Table::new_transaction](crate::table::Table::new_transaction) to create a transaction from
+    /// a table automatically backed by the latest snapshot.
+    pub(crate) fn new(snapshot: impl Into<Arc<Snapshot>>) -> Self {
+        Transaction {
+            read_snapshot: snapshot.into(),
+            operation: None,
+            commit_info: None,
+        }
+    }
+
+    /// Consume the transaction and commit it to the table. The result is a [CommitResult] which
+    /// will include the failed transaction in case of a conflict so the user can retry.
+    pub fn commit(self, engine: &dyn Engine) -> DeltaResult<CommitResult> {
+        // step one: construct the iterator of actions we want to commit
+        // note: only support commit_info right now (and it's required)
+        let engine_commit_info = self
+            .commit_info
+            .as_ref()
+            .ok_or_else(|| Error::MissingCommitInfo)?;
+        let actions = Box::new(iter::once(generate_commit_info(
+            engine,
+            self.operation.as_deref(),
+            engine_commit_info.as_ref(),
+        )?));
+
+        // step two: set new commit version (current_version + 1) and path to write
+        let commit_version = self.read_snapshot.version() + 1;
+        let commit_path =
+            ParsedLogPath::new_commit(self.read_snapshot.table_root(), commit_version)?;
+
+        // step three: commit the actions as a json file in the log
+        let json_handler = engine.get_json_handler();
+        match json_handler.write_json_file(&commit_path.location, Box::new(actions), false) {
+            Ok(()) => Ok(CommitResult::Committed(commit_version)),
+            Err(Error::FileAlreadyExists(_)) => Ok(CommitResult::Conflict(self, commit_version)),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Set the operation that this transaction is performing. This string will be persisted in the
+    /// commit and visible to anyone who describes the table history.
+    pub fn operation(&mut self, operation: String) {
```

Comment:
nit: Isn't `with_xxx` and consuming `self` the more typical approach?
```suggestion
        pub fn with_operation(self, operation: String) {
```
(again below)

Reply:
Yup, both of operation and commit info are candidates for a builder pattern like you mentioned (but would need to return `Self`) which means they could also probably be parameters to `Transaction::new`... I suppose this is worth tackling now so we don't end up introducing APIs we don't like. How about the 'builder option 1' below? This allows us to have a clear builder pattern for the transaction and still have the notion of a `Transaction` which will soon need APIs to fetch write_context and add parquet metadata which wouldn't really fit into the builder pattern in the 'builder option 2'. wdyt @nicklan @scovich 

```rust
// existing
let txn = Transaction::new(...); // do work to create snapshot
txn.operation("insert");
txn.commit_info(something);
txn.commit(); // consume Transaction

// builder option 1 [preferred]: build a Transaction
let builder = TransactionBuilder::new(...);
let txn: Transaction = builder
    .with_operation("insert")
    .with_commit_info(something)
    .build(); // do work to create snapshot
txn.commit(); // consume Transaction

// builder option 2: Transaction itself is a builder
let txn = Transaction::new(...) // do the work to create snapshot here?
    .with_operation("insert")
    .with_commit_info(something);
txn.commit(); // consume Transaction
```

Reply:
Had some discussion with @nicklan and we are landing on option 2 (which is basically what @scovich mentions above). Considering we are designing a bit for a 'future API' that we don't know yet, I'm going to make the call to move forward here and if we have to convert to builder in option 1 etc. then we can :)